### ISSUE TITLE
Python与机器学习实战：从数据预处理到模型部署

### ISSUE THREAD

**Issue by sbumsid24** [sbumsid24/TG#4]
在当今数据驱动的世界中，机器学习（Machine Learning, ML）已成为计算机科学的核心技术之一。Python凭借其丰富的库（如NumPy、Pandas、Scikit-learn、TensorFlow）成为机器学习的首选语言。本文将详细介绍如何使用Python进行完整的机器学习项目开发，涵盖数据预处理、模型训练、评估及部署等关键步骤，并提供大量可运行的代码示例。

2. 数据预处理
2.1 加载和探索数据
使用Pandas加载CSV数据并进行分析：

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# 加载数据集
data = pd.read_csv("data.csv")
print(data.head())  # 查看前5行数据
print(data.info())  # 查看数据信息
print(data.describe())  # 统计描述

# 可视化数据分布
data["target"].value_counts().plot(kind="bar")
plt.title("Class Distribution")
plt.show()
2.2 数据清洗
处理缺失值和异常值：

# 检查缺失值
print(data.isnull().sum())

# 填充缺失值
data.fillna(data.mean(), inplace=True)  # 数值型数据用均值填充
data["category_column"].fillna(data["category_column"].mode()[0], inplace=True)  # 分类数据用众数填充

# 去除异常值
Q1 = data["numeric_column"].quantile(0.25)
Q3 = data["numeric_column"].quantile(0.75)
IQR = Q3 - Q1
data = data[(data["numeric_column"] >= Q1 - 1.5*IQR) & (data["numeric_column"] <= Q3 + 1.5*IQR)]
2.3 特征工程
标准化和编码分类变量：

from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer

# 定义数值和分类列
numeric_features = ["age", "income"]
categorical_features = ["gender", "education"]

# 构建预处理管道
preprocessor = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), numeric_features),
        ("cat", OneHotEncoder(), categorical_features)
    ]
)

# 应用预处理
X = preprocessor.fit_transform(data.drop("target", axis=1))
y = data["target"]
3. 模型训练与评估
3.1 训练-测试集划分
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
3.2 训练随机森林模型
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# 初始化模型
model = RandomForestClassifier(n_estimators=100, random_state=42)

# 训练模型
model.fit(X_train, y_train)

# 预测
y_pred = model.predict(X_test)

# 评估模型
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))
3.3 超参数调优（GridSearchCV）
from sklearn.model_selection import GridSearchCV

param_grid = {
    "n_estimators": [50, 100, 200],
    "max_depth": [None, 10, 20],
    "min_samples_split": [2, 5, 10]
}

grid_search = GridSearchCV(RandomForestClassifier(), param_grid, cv=5)
grid_search.fit(X_train, y_train)

print("Best Parameters:", grid_search.best_params_)
print("Best Score:", grid_search.best_score_)
4. 深度学习模型（TensorFlow/Keras）
4.1 构建神经网络
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

# 构建模型
model = Sequential([
    Dense(128, activation="relu", input_shape=(X_train.shape[1],)),
    Dropout(0.3),
    Dense(64, activation="relu"),
    Dropout(0.3),
    Dense(1, activation="sigmoid")  # 二分类输出
])

# 编译模型
model.compile(optimizer="adam", loss="binary_crossentropy", metrics=["accuracy"])

# 训练模型
history = model.fit(X_train, y_train, epochs=20, batch_size=32, validation_split=0.2)

# 评估模型
loss, accuracy = model.evaluate(X_test, y_test)
print(f"Test Accuracy: {accuracy*100:.2f}%")
4.2 模型可视化
# 绘制训练和验证损失
plt.plot(history.history["loss"], label="Training Loss")
plt.plot(history.history["val_loss"], label="Validation Loss")
plt.legend()
plt.show()

# 绘制训练和验证准确率
plt.plot(history.history["accuracy"], label="Training Accuracy")
plt.plot(history.history["val_accuracy"], label="Validation Accuracy")
plt.legend()
plt.show()
5. 模型部署（Flask API）
5.1 保存模型
import joblib

# 保存Scikit-learn模型
joblib.dump(model, "random_forest_model.pkl")

# 保存Keras模型
model.save("keras_model.h5")
5.2 创建Flask API
from flask import Flask, request, jsonify
import joblib
import numpy as np

app = Flask(__name__)

# 加载模型
model = joblib.load("random_forest_model.pkl")

@app.route("/predict", methods=["POST"])
def predict():
    data = request.get_json()
    features = np.array(data["features"]).reshape(1, -1)
    prediction = model.predict(features)
    return jsonify({"prediction": int(prediction[0])})

if __name__ == "__main__":
    app.run(host="0.0.0.0", 
https://yijingbazi.com/baike/2324.html
https://yijingbazi.com/baike/2323.html
https://yijingbazi.com/baike/2322.html
https://yijingbazi.com/baike/2321.html
https://qyqgw.com/baike/3128.html
https://qyqgw.com/baike/3127.html
https://qyqgw.com/baike/3086.html
https://qyqgw.com/baike/3085.html
https://qyqgw.com/baike/3084.html
https://qyqgw.com/baike/3083.html
https://hunyinwenti.com/baike/2833.html
https://hunyinwenti.com/baike/2832.html
https://hunyinwenti.com/baike/2831.html
https://hunyinwenti.com/baike/2830.html
https://hunyinwenti.com/baike/2829.html
https://hunyinwenti.com/baike/2828.html
https://yijingbazi.com/baike/2382.html
https://yijingbazi.com/baike/2381.html
https://yijingbazi.com/baike/2380.html
